### ISSUE TITLE
Linter destroys performance on large files

### ISSUE THREAD

**Issue by toomuchpete** [atom-community/linter#708]
Had to disable linter (and probably go back to Sublime Text) because Linter is causing the editor to take 1-2s to respond to keystrokes moving through a file. It's an 9,000 line file with a bunch of linting errors (1,700+) in it. 

I'm running three individual linters (`linter-php`, `linter-phpcs`, and `linter-phpmd`) if any of them are turned on, the app grinds to a halt. Seems like there might be a problem with the way the larger linter plugin handles and displays large number of linting errors.

**Comment by steelbrain** [comment]
You have `1700` errors in one file, Woot! :sweat_smile: 
Rest assured we are working on making linter more efficient and faster, we already have a lot of performance improvements in the master branch and will be released soon.
Until then, You can just disable linting for that one particular file while still using it all over your project, by focus that text editor and running `linter:toggle` from the command pallete.

**Comment by toomuchpete** [comment]
@steelbrain Thanks for the quick reply! I'll keep an eye on future releases, but not having access to it for the bigger files is pretty much a deal breaker for me right now, so back to ST I go.

**Comment by steelbrain** [comment]
While profiling today, I set up a few hooks in the DOM updates and realized that we are doing stuff really inefficiently, I have some updates planned that should fix them.

**Comment by nemoDreamer** [comment]
:+1: could we have a setting for the throttle/debounce threshold? i really don't mind waiting 1s for linter to kick in if it means it's not continuously linting while i type (which combined w/ the hovering tooltip can really get in the way of seeing your context)

`linter:toggle` is per editor? that would also help with my "too many errors / panel hiding entire editor" gripe in #722

**Comment by steelbrain** [comment]
We don't really keep a threshold in our codebase, we use the core's `onDidStopChanging` event. and we've found it to work great, if you have any performance issues with linters, we would recommend disabling `lintOnFly` so you only get results when you save a file.

**Comment by nemoDreamer** [comment]
Yeah, I worked w/ that for a while, but it gets confusing fast when all the markings don't move w/ the line-changes, etc. But yes, it's definitely a valid workaround.

Could I change my core's `onDidStopChanging` threshold? Maybe there's a plug-in for that :wink:, will see before I start muckin'about.

**Comment by steelbrain** [comment]
It's hard-coded I think.

**Comment by mark-bradshaw** [comment]
I'd appreciate having an option to quickly disable linting via a keyboard shortcut.  I sometimes need to move into files that are dependencies of my projects, and I don't want the linter jogging my elbow and slowing things down the whole time.  A quick keyboard shut down while I'm in there would be handy.

**Comment by steelbrain** [comment]
@mark-bradshaw you can always associate commands with keybindings of your choice in your `keymap.cson` of atom

**Comment by druellan** [comment]
Yeah, I can confirm it is a problem on large files. I'm currently working on a very big 8.000 lines CSS file, and Linter freezes the editor for about 2 seconds, even with csslint disabled.
Something I notice is that the file is linted on every keystroke, and I believe that some kind of delay, and/or exec interruption could useful to avoid saturate the editor.

**Comment by AsaAyers** [comment]
The linter is delayed. We use atoms built in mechanism that has a 300ms
delay. If that isn't long enough you can open an issue in atom/atom

**Comment by druellan** [comment]
Thanks!
Mmmm, interesting. I can confirm this is not the case here, at least, Atom freezes at the moment I push the first key, and the time depends on the keystrokes.
For example, if I type "d", Atom freezes for about 2 seconds before the "d" character is displayed; if I type "delayed" very fast, Atom freezes for about 10 seconds before the whole word is displayed. CPU usage gets to 30% (i5 4200U CPU).
All community packages disabled except Linter.
@toomuchpete can you please confirm that?

**Comment by toomuchpete** [comment]
@druellan I recall that also being the behavior I experienced, but I'll check again when up-to-date versions of everything.

**Comment by toomuchpete** [comment]
So here's what I discovered. I'm running Atom 1.0.2 and Linter 1.2.4

When running JUST the linter core but no actual linter, I can't tell if it is linting on every keystroke or not, there's no noticeable lag.

When I enable linter-php (v1.0.2), it doesn't lint on every keystroke, but the pause required is VERY short, probably on the order of 50ms rather than 300. I'm not sure how to measure it more precisely, but I know if I create a new function in my 9000 line file and type: `this` followed by a bunch of spaces, I can keep it from showing the syntax error indefinitely, but if I type at anything other than a very brisk pace, it will lint while I'm typing. Any small pause gives it the window it needs. 

That said, it doesn't introduce any sort of visible lag or lock-up.

When I turn on the other linters (PHPCS and PHPMD) I get similar behavior although the lag is more noticeable.

I think if the pause requirement were closer to the 300ms it would be much less of an issue.

**Comment by Arcanemagus** [comment]
PHPMD is still using the old API (Waiting on https://github.com/AtomLinter/linter-phpmd/pull/11), which I believe has a much higher perfomance/lag hit than linters utilizing the new API.

That said, the wait time before linting for lintOnFly I would say is way too short, I also run into issues with large files slowing things to a crawl.

**Comment by druellan** [comment]
I'm going to try again removing the Atom config altogether, and the Linter core package only.
BTW, this is the file I'm working on:
http://tempsend.com/3BEDCE2824

**Comment by druellan** [comment]
I can confirm that Linter core alone does not introduce any lag. But Linter + linter-csslint on the above file, introduces 2 sec lag on every keypress.

**Comment by steelbrain** [comment]
https://github.com/atom-community/linter/pull/787 has provided the base to make linter efficient, we are using efficient markers now, what's left is bubble and bottom panel.

**Comment by druellan** [comment]
Thanks for the follow up @steelbrain

**Comment by steelbrain** [comment]
To all those who have subscribed, I have a diff up in https://github.com/atom-community/linter/pull/827 which introduces super-efficient linter panel updates, which means that you can work painlessly on as many messages as your ram can bear (tested upto 5k) without any kind-of performance regression.
The only thing that's left is bubble, You might have to disable it if you want superior performance. but You can expect a fix for it too, soon.

**Comment by aseemk** [comment]
Hi folks, I just updated to the latest Linter and linter-coffeelint (I wish I remembered which versions I was on previously, but they were both at least several months old).

I've turned off "Lint on change" and "Lint on the fly", so it only lints on saves. Unfortunately, I too am seeing a massive performance regression: every time I save, Atom locks up for 2 secs or so.

My current example is a ~5k LOC CoffeeScript file with no errors or warnings. I hope that's a helpful example, where I'd expect the bottleneck not to be UI updates, but something else (e.g. the linting).

(And I'm ignorant here, but if it's not UI updates, would it be possible to run any processing JS in a separate thread/worker, so it doesn't lock up the UI?)

**Comment by steelbrain** [comment]
@aseemk can you try that PR and report the diff?

**Comment by aseemk** [comment]
Sure, can you let me know how to do that? (Atom dev noob here.) Thanks. =)

**Comment by steelbrain** [comment]
@aseemk 
```
mv ~/.atom/packages/linter /tmp/linter-old
cd ~/.atom/packages
git clone https://github.com/atom-community/linter
cd linter
apm install
```

**Comment by Arcanemagus** [comment]
Since that PR isn't merged into master you would also need to run `git checkout steelbrain/rename-views-ui` before `apm install`. Otherwise you will be simply running the latest master.

**Comment by aseemk** [comment]
@steelbrain: Sorry I didn't get to try the PR out! Glad to see it's merged though. Looking forward to trying it out on the next update. =)

**Comment by steelbrain** [comment]
@aseemk a release is out :)

**Comment by aseemk** [comment]
Sweet, thanks! I just updated and it seems to be noticeably faster indeed! =)

**Comment by druellan** [comment]
Thanks! Yes, works really nice now =)

**Comment by neoziro** [comment]
Problem is happening again, I opened devtools an tried to investigate:

<img width="517" alt="screen shot 2017-09-17 at 20 16 57" src="https://user-images.githubusercontent.com/266302/30523628-2f72e090-9be5-11e7-85c7-2839b20075a8.png">

The problem only happens when "Show decorations" is active.

**Comment by Arcanemagus** [comment]
@neoziro `linter` works _completely_ differently from when this issue was filed, as a major point it doesn't even have anything to do with display of the messages, that's handled by whatever UI you have installed (`linter-ui-default` being installed automatically for you).

Please file a new issue there if you think this is something to do with `linter-ui-default`, but this could just be an issue with Atom as I have seen similar issues reported there with large numbers of decorations.

Locking this issue as it has no bearing on the current state of this package.
